include/matx/kernels/conv.cuh

```diff
@@ -187,91 +188,90 @@ __global__ void Conv1D(OutType d_out, InType d_in, FilterType d_filter,
+                       matxConvCorrMode_t mode, int num_batch)
 {
-  extern __shared__ float s_exch[];
-  using ftype_strip = typename FilterType::scalar_type;
-  using intype_strip = typename InType::scalar_type;
-  using outtype_strip = typename OutType::scalar_type;
-  [[maybe_unused]] index_t bdims[2];
-  ftype_strip *s_filter = reinterpret_cast<ftype_strip *>(&s_exch[0]);
-  int tid_x = blockDim.x * blockIdx.x + threadIdx.x;
-  int tid_y = blockDim.y * blockIdx.y + threadIdx.y;
-  int ix_size, iy_size;
-  // For rank 4 we need to break up the batch dimensions into the two outer
-  // components. Other ranks store the batch index directly in the grid
-  // index
-  if constexpr (d_in.Rank() == 4) {
-    bdims[0] = blockIdx.z / d_in.Size(1);
-    bdims[1] = blockIdx.z - (bdims[0] * d_in.Size(1));
-    ix_size = d_in.Size(3);
-    iy_size = d_in.Size(2);
-  }
-  if constexpr (d_in.Rank() == 3) {
-    ix_size = d_in.Size(2);
-    iy_size = d_in.Size(1);
-  }
-  else if constexpr (d_in.Rank() == 2) {
-    ix_size = d_in.Size(1);
-    iy_size = d_in.Size(0);
-  }
+  using in1type = typename InType1::scalar_type;
+  using in2type = typename InType1::scalar_type;
+  using outtype = typename OutType::scalar_type;
 
-  if ((threadIdx.x < d_filter.Size(1)) && (threadIdx.y < d_filter.Size(0))) {
-    s_filter[d_filter.Size(1) * threadIdx.y + threadIdx.x] =
-        d_filter(threadIdx.y, threadIdx.x);
-  }
+  constexpr int Rank = OutType::Rank();
 
-  __syncthreads();
+  index_t oN = d_out.Size(Rank-2);
+  index_t oM = d_out.Size(Rank-1);
 
-  outtype_strip val = 0;
+  index_t i1N = d_in1.Size(Rank-2);
+  index_t i1M = d_in1.Size(Rank-1);
 
-  for (int x = 0; x < d_filter.Size(1); x++) {
-    if ((tid_x - static_cast<int>(d_filter.Size(1)) + 1 + x < 0) ||
-        (tid_x - static_cast<int>(d_filter.Size(1)) + 1 + x >= ix_size)) {
-      continue;
-    }
+  index_t i2N = d_in2.Size(Rank-2);
+  index_t i2M = d_in2.Size(Rank-1);
 
-    for (int y = 0; y < d_filter.Size(0); y++) {
-      if ((tid_y - static_cast<int>(d_filter.Size(0)) + 1 + y < 0) ||
-          (tid_y - static_cast<int>(d_filter.Size(0)) + 1 + y >= iy_size)) {
-        continue;
-      }
+  int dy = 0, dx = 0;
 
-      if constexpr (d_in.Rank() == 4) {
-      val = detail::madd(s_filter[y * d_filter.Size(1) + x],
-               d_in(bdims[0], bdims[1], tid_y - d_filter.Size(0) + 1 + y,
-                    tid_x - d_filter.Size(1) + 1 + x), val);
-      }
-      else if constexpr (d_in.Rank() == 3) {
-        val = detail::madd(s_filter[y * d_filter.Size(1) + x],
-               d_in(blockIdx.z, tid_y - d_filter.Size(0) + 1 + y,
-                    tid_x - d_filter.Size(1) + 1 + x), val);
-      }
-      else if constexpr (d_in.Rank() == 2) {
-        val = detail::madd(s_filter[y * d_filter.Size(1) + x],
-               d_in(tid_y - d_filter.Size(0) + 1 + y,
-                    tid_x - d_filter.Size(1) + 1 + x), val);
-      }
-    }
-  }
+  // TODO detect mode based on output size?
```

Comment:
I think it's better to be explicit how you have it and just error if the output size is wrong. it gives us flexibility in the future to do dynamic tensor allocations